internal/venice-common/src/main/java/com/linkedin/venice/controller/transport/ControllerGrpcTransportClient.java

```diff
@@ -0,0 +1,103 @@
+package com.linkedin.venice.controller.transport;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.linkedin.venice.client.exceptions.VeniceClientException;
+import com.linkedin.venice.controller.converters.GrpcConvertersRegistry;
+import com.linkedin.venice.controller.converters.RequestConverter;
+import com.linkedin.venice.controller.converters.ResponseConverter;
+import com.linkedin.venice.controller.requests.ControllerRequest;
+import com.linkedin.venice.controllerapi.ControllerResponse;
+import com.linkedin.venice.grpc.GrpcUtils;
+import com.linkedin.venice.protocols.VeniceControllerGrpcServiceGrpc;
+import com.linkedin.venice.security.SSLFactory;
+import io.grpc.ChannelCredentials;
+import io.grpc.Grpc;
+import io.grpc.InsecureChannelCredentials;
+import io.grpc.ManagedChannel;
+import io.grpc.TlsChannelCredentials;
+import java.lang.reflect.Method;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+
+/**
+ * Transport client implementation to interact with gRPC server. The current implementation keeps the lifecycle
+ * of the client at a request granularity as opposed to a longer client due to its usage pattern across the codebase.
+ * Consider expanding the lifecycle to per component or per JVM as we sit with evolution.
+ */
+public class ControllerGrpcTransportClient implements AutoCloseable {
+  private static final long SHUTDOWN_TIMEOUT = 10000L;
+  private final ManagedChannel channel;
+
+  private final GrpcConvertersRegistry converterRegistry;
+
+  public ControllerGrpcTransportClient(String server) {
+    this(server, InsecureChannelCredentials.create());
+  }
+
+  public ControllerGrpcTransportClient(String server, SSLFactory sslFactory) {
+    this(server, buildChannelCredentials(sslFactory));
+  }
+
+  private ControllerGrpcTransportClient(String server, ChannelCredentials channelCredentials) {
+    this.channel = Grpc.newChannelBuilder(server, channelCredentials).build();
+    this.converterRegistry = new GrpcConvertersRegistry();
+  }
+
+  public <REQ extends ControllerRequest, RES extends ControllerResponse> RES request(REQ request)
+      throws ExecutionException, TimeoutException {
+
+    GrpcRoute grpcRoute = RouteUtils.getGrpcRouteFor(request.getRoute());
+
+    try {
+      VeniceControllerGrpcServiceGrpc.VeniceControllerGrpcServiceBlockingStub stub = buildStub();
+
+      Method method = stub.getClass().getMethod(grpcRoute.getMethod(), grpcRoute.getRequestType());
+      Object responseObject = method.invoke(stub, convertToGrpcRequest(request));
```

Comment:
Must we use reflection? Is here not a design pattern that could allow us to plug in the correct stuff without it? e.g. perhaps this "stub" class can return a `Function<REQ, RES>`, or similar?

Reply:
Good point. Of the 5 days in the hackathon, I spent considerable amount of time to make this generic instead of using reflection. Most of my attempts failed due to type erasure and the need to introduce type token to carry information in a way it could potentially work. 

That said, GRPC way of defining these methods are also sort of either too abstracted or too narrow; e.g., the stub generated with createStore method takes a concrete CreateStoreRequest so having type bounds on `REQ` and `RES` to work seemed a bit more tricky and hence I chose this one for now.

I will continue looking into this and see if `TypeToken` is the way to go and perhaps refactor this out if I land on something promising.

Reply:
Ok well, this is not a hot path so the performance downside of reflection is not a concern... the only concern is the maintainability aspect since we lost strong typing. I guess it could be fine to let this in as is, and keep iterating later.

Reply:
Personally, I’m not a fan of using GrpcConvertersRegistry solely to avoid little boilerplate code. I think having some boilerplate is fine as long as it’s simple, easy to maintain, and easy to test.

Reply:
I believe this could be really be done something like:

```java

// ControllerTransport Interface
public interface ControllerTransport {
    ClusterDiscoveryResponse discoverCluster(ClusterDiscoveryRequest storeName);
    LeaderControllerResponse getLeaderController(GetLeaderControllerRequest clusterName);
}

// ControllerGrpcTransport Implementation
public class ControllerGrpcTransport implements ControllerTransport {
    @Override
    public ClusterDiscoveryResponse discoverCluster(ClusterDiscoveryRequest request) {
        System.out.println("gRPC implementation for discoverCluster");
        // Implementation goes here
        return new ClusterDiscoveryResponse("gRPC: Cluster for store " + request.getStoreName());
    }

    @Override
    public LeaderControllerResponse getLeaderController(GetLeaderControllerRequest request) {
        System.out.println("gRPC implementation for getLeaderController");
        // Implementation goes here
        return new LeaderControllerResponse("gRPC: Leader for cluster " + request.getClusterName());
    }
}

// ControllerHttpTransport Implementation
public class ControllerHttpTransport implements ControllerTransport {
    @Override
    public ClusterDiscoveryResponse discoverCluster(ClusterDiscoveryRequest request) {
        System.out.println("HTTP implementation for discoverCluster");
        // Implementation goes here
        return new ClusterDiscoveryResponse("HTTP: Cluster for store " + request.getStoreName());
    }

    @Override
    public LeaderControllerResponse getLeaderController(GetLeaderControllerRequest request) {
        System.out.println("HTTP implementation for getLeaderController");
        // Implementation goes here
        return new LeaderControllerResponse("HTTP: Leader for cluster " + request.getClusterName());
    }
}
```